src/compiler_pruner.cpp

```diff
@@ -22,17 +22,58 @@ const IR::P4Program *apply_unused_decls(const IR::P4Program *program,
         new P4::CreateBuiltins(), new P4::ResolveReferences(&refMap, true),
         new P4::ConstantFolding(&refMap, nullptr),
         new P4::InstantiateDirectCalls(&refMap),
-        new P4::TypeInference(&refMap, &typeMap, false),
-        new ExtendedUnusedDeclarations(&refMap)};
+        new P4::TypeInference(&refMap, &typeMap, false)};
 
-    INFO("Applying custom RemoveAllUnusedDeclarations...");
+    INFO("Applying Generic passes...");
     temp = program->apply(pass_manager);
     emit_p4_program(temp, pruner_conf.out_file_name);
+    if (check_pruned_program(&program, temp, pruner_conf) == EXIT_SUCCESS) {
+        *applied = true;
+    }
+
+    return program;
+}
+
+const IR::P4Program *apply_replace_vars(const IR::P4Program *program,
+                                        P4PRUNER::PrunerConfig pruner_conf,
+                                        bool genericPassesApplied) {
+    INFO("Replacing variables...");
+    const IR::P4Program *temp =
+        replace_variables(program, pruner_conf, genericPassesApplied);
+    emit_p4_program(temp, pruner_conf.out_file_name);
     check_pruned_program(&program, temp, pruner_conf);
 
     return program;
 }
 
+const IR::P4Program *apply_unused_decls(const IR::P4Program *program,
+                                        P4PRUNER::PrunerConfig pruner_conf,
+                                        bool applied) {
+    P4::ReferenceMap refMap;
+    P4::TypeMap typeMap;
+    const IR::P4Program *temp;
+
+    std::initializer_list<Visitor *> passes;
+    PassManager pass_manager(passes);
+
+    if (not applied) {
+        pass_manager.addPasses({new P4::CreateBuiltins(),
+                                new P4::ResolveReferences(&refMap, true),
+                                new P4::ConstantFolding(&refMap, nullptr),
+                                new P4::InstantiateDirectCalls(&refMap),
+                                new P4::TypeInference(&refMap, &typeMap, false),
+                                new ExtendedUnusedDeclarations(&refMap)});
```

Comment:
Just realized that ExtendedUnusedDeclarations only uses the refMap. So we might get away with removing some of the other passes preceding it.

Reply:
Now that we are just applying the generic passes once, it doesn't matter right? Since, all these are required for replacing variables right?

Reply:
Yeah that is also okay